content/academy/api_scraping/general_api_scraping/handling_pagination.md

```diff
@@ -0,0 +1,205 @@
+    const page = await browser.newPage();
+
+    let clientId = null;
+
+    page.on('response', async (res) => {
+        const id = new URL(res._url).searchParams.get('client_id') ?? null;
+        if (id) clientId = id;
+    });
+
+    await page.goto('https://soundcloud.com/tiesto/tracks');
+    await page.waitForSelector('.sc-classic');
+    await browser.close();
+
+    // return the client_id
+    return clientId;
+};
+
+// export the function to be used in a different file
+module.exports = scrapeClientId;
+```
+
+Now, in a new file called **index.js** we'll write the skeleton for our pagination and item-scraping code:
+
+```JavaScript
+// index.js
+const scrapeClientId = require('./scrapeClientId');
+// we will need gotScraping to make HTTP requests
+const { gotScraping } = require('got-scraping');
+
+const scrape100Items = async () => {
+    // the initial request URL
+    let nextHref = 'https://api-v2.soundcloud.com/users/141707/tracks?limit=20&offset=0';
+
+    // create an array for all of our scraped items to live
+    let items = [];
+
+    // scrape the client ID with the script from the
+    // previous lesson
+    const clientId = await scrapeClientId();
+
+    // More code will go here
+};
+```
+
+Let's now take a step back and think about the condition on which we should continue paginating:
+
+1. If the API responds with a **next_href** set to **null**, that means that there are no more pages, and that we have scraped all of the possible items and we should stop paginating.
+2. If our items list has 100 records or more, we should stop paginating. Otherwise, we should continue until 100+ items has been reached.
+
+With a full understanding of this condition, we can translate it into code:
+
+```JavaScript
+// continue making requests until either we've reached 100+ items
+while (items.flat().length < 100) {
+    // if the "next_href" wasn't present in the last call, there
+    // are no more pages. return what we have and stop paginating.
+    if (!nextHref) return items.flat();
+
+    // continue paginating
+}
+```
+
+All that's left to do now is flesh out this `while` loop with pagination logic and finally return the **items** array once the loop has finished.:
+
+```JavaScript
+// index.js
+const scrapeClientId = require('./scrapeClientId');
+const { gotScraping } = require('got-scraping');
+
+const scrape100Items = async () => {
+    let nextHref = 'https://api-v2.soundcloud.com/users/141707/tracks?limit=20&offset=0';
+    let items = [];
+
+    const clientId = await scrapeClientId();
+
+    while (items.flat().length < 100) {
+        if (!nextHref) return items.flat();
+
+        // set the "client_id" URL parameter of the
+        // nextHref URL
+        const nextURL = new URL(nextHref);
+        nextURL.searchParams.set('client_id', clientId);
+
+        // make the paginated request and push its results
+        // into the in-memory "items" array
+        const res = await gotScraping(nextURL.toString());
```

Comment:
I would at least mention that it is better to add requests to the queue instead of processing them in memory and that Apify Crawlers give you that logic